libs/constrained-generators/src/Constrained/Base.hs

```diff
@@ -3715,3 +3769,378 @@ instance HasSpec fn a => Pretty (Spec fn a) where
+sizeOfFn = injectFn $ SizeOf @fn @a
+
+-- Operations on Size (specified in SizeFn) by the Functions instance
+
+instance (BaseUniverse fn, HasSpec fn Size) => Functions (SizeFn fn) fn where
+  propagateSpecFun _ _ TrueSpec = TrueSpec
+  propagateSpecFun _ _ (ErrorSpec err) = ErrorSpec err
+  propagateSpecFun fn (ListCtx pre HOLE suf) (SuspendedSpec x p) =
+    constrained $ \x' ->
+      let args = appendList (mapList (\(Value a) -> Lit a) pre) (x' :> mapList (\(Value a) -> Lit a) suf)
+       in Let (App (injectFn fn) args) (x :-> p)
+  propagateSpecFun SizeOf (NilCtx HOLE) (TypeSpec x _) = liftSizeSpec x
+  propagateSpecFun SizeOf (NilCtx HOLE) (MemberSpec xs) = liftMemberSpec xs
+
+  mapTypeSpec f ts = mapTypeSpecSize f ts
+
+mapTypeSpecSize :: forall fn a b f. f ~ SizeFn fn => f '[a] b -> TypeSpec fn a -> Spec fn b
+mapTypeSpecSize f ts = case f of
+  SizeOf ->
+    constrained $ \x ->
+      unsafeExists $ \x' ->
+        assert (x ==. sizeOf_ x') <> toPreds @fn @a x' ts
+
+-- ======================================
+type SizeSpec = NumSpec Integer
+
+-- Operations that build SizeSpec from Integer, but raise an error if called on a negative numbers
+exactSize :: Size -> SizeSpec
+exactSize a
+  | a >= 0 = rangeSize a a
+  | otherwise = error ("Negative Int in call to exactSize: " ++ show a)
+
+rangeSize :: Size -> Integer -> SizeSpec
+rangeSize a b | a < 0 || b < 0 = error ("Negative Int in call to rangeSize: " ++ show a ++ " " ++ show b)
+rangeSize a b = NumSpecInterval (Just a) (Just b)
+
+atLeastSize :: Size -> SizeSpec
+atLeastSize a
+  | a < 0 = error ("Negative Int in call to atLeastSize: " ++ show a)
+  | otherwise = NumSpecInterval (Just a) Nothing
+
+atMostSize :: Size -> SizeSpec
+atMostSize a
+  | a < 0 = error ("Negative Int in call to atMostSize: " ++ show a)
+  | otherwise = NumSpecInterval Nothing (Just a)
+
+exactSizeSpec :: BaseUniverse fn => Size -> Spec fn Size
+exactSizeSpec a = MemberSpec [a]
+
+atMostSpec :: BaseUniverse fn => Spec fn Size -> Spec fn Size
+atMostSpec TrueSpec = TrueSpec
+atMostSpec (SuspendedSpec _ _) = TrueSpec
+atMostSpec (ErrorSpec xs) = ErrorSpec xs
+atMostSpec (MemberSpec []) = ErrorSpec ["empty MemberSec in atMostSpec."]
+atMostSpec (MemberSpec xs) = typeSpec (atMostSize (maximum xs))
+atMostSpec (TypeSpec (NumSpecInterval _ hi) bad) = TypeSpec (NumSpecInterval Nothing hi) bad
+
+-- ================
+-- Sized
+-- ================
+
+class Sized t where
+  sizeOf :: t -> Size
+  liftSizeSpec :: HasSpec fn t => SizeSpec -> Spec fn t
+  liftMemberSpec :: HasSpec fn t => OrdSet Size -> Spec fn t
+  sizeOfTypeSpec :: HasSpec fn t => TypeSpec fn t -> Spec fn Size
+
+instance Sized Size where
+  sizeOf x = x
+  liftSizeSpec x = typeSpec x
+  liftMemberSpec xs = MemberSpec xs
+  sizeOfTypeSpec x = (TypeSpec x [])
+
+instance Sized Int where
+  sizeOf x = fromIntegral x
+  liftSizeSpec (NumSpecInterval a b) = typeSpec (NumSpecInterval (fromIntegral <$> a) (fromIntegral <$> b))
```

Comment:
The uses of fromIntegral here are not safe w.r.t max and min int.

Reply:
we got rid of these 2 instances.